src/platform/w3dengine/client/w3ddisplay.cpp

```diff
@@ -0,0 +1,950 @@
+    m_3DInterfaceScene = nullptr;
+    m_averageFps = g_theWriteableGlobalData->m_framesPerSecondLimit;
+    for (auto &str : m_displayStrings) {
+        str = nullptr;
+    }
+    for (auto &light : m_myLight) {
+        light = nullptr;
+    }
+}
+
+// 0x0073C453
+W3DDisplay::~W3DDisplay()
+{
+    //delete m_debugDisplay;
+    //for (auto &str : m_displayStrings) {
+    //    g_theDisplayStringManager->Free_Display_String(str);
+    //}
+
+    //delete m_2DRender;
+    //m_2DRender = nullptr;
+    //Delete_Views();
+    //m_3DScene->Release_Ref();
+    //m_3DScene = nullptr;
+    //m_2DScene->Release_Ref();
+    //m_2DScene = nullptr;
+    //m_3DInterfaceScene->Release_Ref();
+    //m_3DInterfaceScene = nullptr;
+
+    //for (auto &light : m_myLight) {
+    //    light->Release_Ref();
+    //    light = nullptr;
+    //}
+
+    //PredictiveLODOptimizerClass::Free();
+    //// unk string 0xA48068 StringClass::Release_Resources();
+    //W3DShaderManager::Shutdown();
+    //m_assetManager->Free_Assets();
+    //delete m_assetManager;
+    //// WW3D::Shutdown();
+    //// WWMath::Shutdown();
+    //// Browser Shutdown();
+    //// delete g_theW3DFileSystem;
+
+    // TODO: Requires WW3D, WWMath, g_theW3DFileSystem, 0xA48068, Browser
+#ifdef GAME_DLL
+    Call_Method<void, W3DDisplay>(0x0073C470, this);
+#endif
+}
+
+// 0x0073CA80
+void W3DDisplay::Init()
+{
+    // TODO: Requires RTS3DInterfaceScene, RTS2DScene, RTS3DScene, W3DAssetManager, ...
+#ifdef GAME_DLL
+    Call_Method<void, W3DDisplay>(0x0073CA80, this);
+#endif
+}
+
+// 0x0073D030
+void W3DDisplay::Reset()
+{
+    auto *iter = m_3DScene->Create_Iterator(false);
+    for (iter->First(); !iter->Is_Done(); iter->Next()) {
+        auto *renderObj = iter->Current_Item();
+        renderObj->Add_Ref();
+        m_3DScene->Remove_Render_Object(renderObj);
+        renderObj->Release_Ref();
+    }
+    m_3DScene->Destroy_Iterator(iter);
+
+    m_isClippedEnabled = false;
+
+    m_assetManager->Release_Unused_Assets();
+
+    g_theWriteableGlobalData->m_drawSkyBox = false;
+}
+
+// 0x0073E360
+void W3DDisplay::Draw()
+{
+    // Requires GameLogic, GameClient, ...
+#ifdef GAME_DLL
+    Call_Method<void, W3DDisplay>(0x0073E360, this);
+#endif
+}
+
+// 0x0073C980
+void W3DDisplay::Set_Width(uint32_t width)
+{
+    Display::Set_Width(width);
+    RectClass rect = { { 0.0f, 0.0f }, { static_cast<float>(Get_Width()), static_cast<float>(Get_Height()) } };
+    m_2DRender->Set_Coordinate_Range(rect);
+}
+
+// 0x0073CA00
+void W3DDisplay::Set_Height(uint32_t height)
+{
+    Display::Set_Height(height);
+    RectClass rect = { { 0.0f, 0.0f }, { static_cast<float>(Get_Width()), static_cast<float>(Get_Height()) } };
+    m_2DRender->Set_Coordinate_Range(rect);
+}
+
+// 0x0073C840
+bool W3DDisplay::Set_Display_Mode(uint32_t width, uint32_t height, uint32_t bits, bool windowed)
+{
+    auto success = DX8Wrapper::Set_Device_Resolution(width, height, bits, windowed, true);
+
+    if (success) {
```

Comment:
if DX8Wrapper::Set_Device_Resolution(width, height, bits, windowed, true) == W3D_ERROR_OK

Reply:
The Set_Device_Resolution function returns a bool are you sure this is correct.

Reply:
I think the issue is this was wrapped by WW3D which modified the return value.

Reply:
Wrapper has been added.